### PR TITLE
RFC: Make hooks sort order locale-independent

### LINKED ISSUE #2494: language: tag is not well-formed
/kind bug

**Description**
Warning about a not-wellformed tag when parsing language.

**Steps to reproduce the issue:**
`podman --log-level debug run -it --rm --read-only busybox sh`

**Describe the results you received:**
`WARN[0000] failed to parse language "en_US.UTF-8": language: tag is not well-formed`

**Describe the results you expected:**
No warning :)

**Additional information you deem important (e.g. issue happens only occasionally):**
```
# locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```

**Output of `podman version`:**
```
podman version 1.2.0-dev
```

**Output of `podman info --debug`:**

```
debug:
  compiler: gc
  git commit: 91373151335a84c5b78dbe46362cb678807d7509
  go version: go1.11.5
  podman version: 1.2.0-dev
host:
  BuildahVersion: 1.7.1
  Conmon:
    package: Unknown
    path: /usr/libexec/podman/conmon
    version: 'conmon version 1.14.0-dev, commit: 8de2ffc9e0809e98a72b93e35b74b8e5d4b0c29c'
  Distribution:
    distribution: debian
    version: "9"
  MemFree: 32108892160
  MemTotal: 38205562880
  OCIRuntime:
    package: Unknown
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc6+dev
      commit: f79e211b1d5763d25fb8debda70a764ca86a0f23
      spec: 1.0.1-dev
  SwapFree: 8586784768
  SwapTotal: 8586784768
  arch: amd64
  cpus: 8
  hostname: pgsql
  kernel: 4.19.0-0.bpo.2-amd64
  os: linux
  rootless: false
  uptime: 26h 48m 2.64s (Approximately 1.08 days)
insecure registries:
  registries: []
registries:
  registries:
  - docker.io
  - registry.fedoraproject.org
  - registry.access.redhat.com
store:
  ConfigFile: /etc/containers/storage.conf
  ContainerStore:
    number: 0
  GraphDriverName: overlay
  GraphOptions: null
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 0
  RunRoot: /var/run/containers/storage
  VolumePath: /var/lib/containers/storage/volumes
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
VirtualBox 6 on Windows 10, running Debian 9 (stable)

### PR DESCRIPTION
**Pull request by mtrmac**
**NOTE:** This is a potentially breaking change.

Don't sort OCI hooks using the locale collation order; it does not make sense for the same system-wide directory to be interpreted differently depending on the user's `LC_COLLATE` setting, and the language-specific collation order can even change over time.
    
Besides, the current collation order determination code has never worked with the most common `LC_COLLATE` values like `en_US.UTF-8`.

Ideally, we would like to just order based on Unicode code points to be reliably stable, but the existing implementation is case-insensitive, so we are forced to rely on the unicode case mapping tables at least.
    
(This gives up on canonicalization and width-insensitivity, potentially breaking users who rely on these previously documented properties.)

Fixes #2494.

Alternatively, we could keep the current implementation and just hard-code the `und-u-va-posix` value to be always used.

**NOTE:** CRI-O contains very similar code, and should be updated similarly if this approach is accepted.

### CODE REVIEW

**Comment by mtrmac** [comment]
@mheon PTAL

**Comment by mheon** [comment]
I recall a discussion in the original PR about why we could do this, but I can't recall any details; I'll look it up and see what our eventual reason for it was.

Code LGTM, though.

File: pkg/hooks/docs/oci-hooks.5.md [hunk 1]
```diff
@@ -23,10 +23,9 @@ If multiple directories are configured, a JSON filename in a preferred directory
...
 
-Hooks are injected in the JSON filename case- and width-insensitive collation order.
-Collation order depends on your locale, as set by `LC_ALL`, `LC_COLLATE`, or `LANG` (in order of decreasing precedence).
-For more information, see `locale(7)`.
-For example, in the POSIX locale, a matching hook defined in `01-my-hook.json` would be injected before matching hooks defined in `02-another-hook.json` and `01-UPPERCASE.json`.
+Hooks are injected in the order obtained by sorting the JSON file names, after converting them to lower case, based on their Unicode codee points.
```
Comment (TomSweeneyRedHat) [inline pkg/hooks/docs/oci-hooks.5.md]: s/codee/code

**Comment by rhatdan** [comment]
@wking Did you put the sort order in originally?

**Comment by wking** [comment]
> @wking Did you put the sort order in originally?

Yeah, previous discussion starting [here][1].

[1]: https://github.com/containers/libpod/pull/686#issuecomment-387910975

**Comment by wking** [comment]
> Besides, the current collation order determination code has never worked with the most common LC_COLLATE values like en_US.UTF-8.

I'd mentioned the difference between languages and locales in golang/go#25340, but haven't followed up beyond that.

**Comment by rhatdan** [comment]
@wking Do you disagree with @mtrmac?

**Comment by wking** [comment]
> Don't sort OCI hooks using the locale collation order; it does not make sense for the same system-wide directory to be interpreted differently depending on the user's `LC_COLLATE` setting, and the language-specific collation order can even change over time.

If you set `LC_COLLATE` (or set `LANG` and leave `LC_COLLATE` unset), you're saying that you want collation to happen in that order.  If you point yourself at a system-wide hook directory where the admin expected a given collation order but you chose another one that sorted the hooks differently, then I'm fine running the hooks in a different order for you.  Conservative admins can use digit prefixes (`00-my-first-hook.yaml`, etc.) that are stable across most (all?) collation locales.  Conservative non-admins can set `--hooks-dir` to point at a hook directory they maintain.  So I'm fine with this aspect of how we currently do things.

We chose locale-based sorting to [avoid byte-for-byte comparison of possibly-unormalized UTF-8][1] (more on normalization [here][2]).  I think avoiding that sort of surprise still makes sense.  It's also low-probability, but I'd rather explain "you set `LC_COLLATE`, so we gave you this order" than explain "ah, you used a combining sequence, while this other hook is using the single, combined character".

> Alternatively, we could keep the current implementation and just hard-code the `und-u-va-posix` value to be always used.

This is effectively what we have now, for any locales that don't parse as languages, right?  I'm ok moving in this direction (it resolves both the "I didn't realize I have a different collation locale" and "I didn't realize Unicode gives you several byte-array implementations for a single semantic string" issues).  But I'm not wild about ignoring the configured collation for users who do intentionally prefer a particular collation.

So I'm not a maintainer, but personally, I'd rather stay away from byte-based sorts and use either the current master code or a hard-coded `und-u-va-posix` that ignores the users preferred collation locale.

[1]: https://github.com/containers/libpod/pull/686#issuecomment-387914358
[2]: http://www.unicode.org/reports/tr15/

**Comment by mtrmac** [comment]
> > Don't sort OCI hooks using the locale collation order; it does not make sense for the same system-wide directory to be interpreted differently depending on the user's `LC_COLLATE` setting, and the language-specific collation order can even change over time.
> 
> If you set `LC_COLLATE` (or set `LANG` and leave `LC_COLLATE` unset), you're saying that you want collation to happen in that order.

Yes, for _ordinary human text_ like dictionaries and address books. It is not trivially applicable to software APIs, just like `LC_NUMERIC` is rarely used to choose between decimal dot/comma in data storage/transports formats (and if it is used, it is almost always an oversight.)

A hooks directory, where ordering has functional impact, is much closer to an API than to a dictionary.

> If you point yourself at a system-wide hook directory where the admin expected a given collation order but you chose another one that sorted the hooks differently, then I'm fine running the hooks in a different order for you.

Even if it _seems to be_ “we are doing the right thing to support different languages and cultures”, this is _in effect_ “we have designed the system so that it breaks because you want your address book sorted correctly for your language, and your system administrator uses a different language”. If ostensibly doing the right thing gets us exactly the opposite result, the assumptions must be challenged.

> Conservative admins can use digit prefixes (`00-my-first-hook.yaml`, etc.) that are stable across most (all?) collation locales.

Yes, quite; but when that’s the _only_ reasonable way to do things, we should just recommend (or enforce?) it instead of (pretending to) support subtly-break-my-system alternatives.

> Conservative non-admins can set `--hooks-dir` to point at a hook directory they maintain.

That’s true, but being forced to maintain an independent configuration just because the software makes it hard to use the system-wide one is not optimal.

> We chose locale-based sorting to [avoid byte-for-byte comparison of possibly-unormalized UTF-8](https://github.com/containers/libpod/pull/686#issuecomment-387914358)

Ah, OK; the original PR talked _only_ about byte-for-byte comparison of the UTF-8 form (which is fine), it did not mention normalization.

> I think avoiding that sort of surprise still makes sense. It's also low-probability, but I'd rather explain "you set `LC_COLLATE`, so we gave you this order" than explain "ah, you used a combining sequence, while this other hook is using the single, combined character".

Honestly my take on _either_ is “you did not start the names with numbers and relied on corner cases, don’t try to be too cute and clever if order matters”, and “we are not including 76k lines of tables in the software product to enable this failure”.

> > Alternatively, we could keep the current implementation and just hard-code the `und-u-va-posix` value to be always used.
> 
> This is effectively what we have now, for any locales that don't parse as languages, right?

I think so, modulo the warnings on every execution :)

> So I'm not a maintainer, but personally, I'd rather stay away from byte-based sorts and use either the current master code or a hard-coded `und-u-va-posix` that ignores the users preferred collation locale.

BTW, what _is_ the definition of that order? I haven’t found anything definitive; meanwhile the POSIX definition of `LC_COLLATE=POSIX` per http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap07.html#tag_07_03_02_06 , and its definition of “character” as “multi-byte character” (which in UTF-8 ends up meaning a single Unicode code point) explicitly forbids making collation normalization-aware, because every code point has its own unique primary weight and that’s that. (Curiously enough, the only references to `und-u-va-posix` that Google finds are all related to this repo and/or conversation.)

**Comment by wking** [comment]
> Even if it seems to be “we are doing the right thing to support different languages and cultures”, this is in effect “we have designed the system so that it breaks because you want your address book sorted correctly for your language, and your system administrator uses a different language”.

... and the admin and user decided to use the same hook directories (high probability) and the admin decided to lead with characters that sort differently in different locales (low probability).  I think that last mitigation is important.

> > Conservative admins can use digit prefixes (`00-my-first-hook.yaml`, etc.) that are stable across most (all?) collation locales.
>
> Yes, quite; but when that’s the *only* reasonable way to do things, we should just recommend (or enforce?)...

Digit prefixes are a reasonable admin approach on multi-user systems with a divergence of preferred collation locales.  That's a ways from "only", so I see no need to enfore.  I'm fine adding recommendations for digit prefixes though.

> ... it instead of (pretending to) support subtly-break-my-system alternatives.

I'm in favor of splitting on strings to get `en_US.UTF-8` -> `en_US` (which does parse correctly) and similar.  I don't think we need to give up on this approach entirely based on that current issue.

> > Conservative non-admins can set `--hooks-dir` to point at a hook directory they maintain.
>
> That’s true, but being forced to maintain an independent configuration just because the software makes it hard to use the system-wide one is not optimal.

Implicit hook directories were [already deprecated in favor of `--hooks-dir`][1] (#1920) before this discussion.  So I don't see this as a big loss.

> ... “we are not including 76k lines of tables in the software product to enable this failure”.

This is also a fair point.  It's not strong enough to tip me over to prefer byte-for-byte comparison though, and I [remain][2] on the fence between the two approaches that depend on the library ;).

> I haven’t found anything definitive; meanwhile the POSIX definition of LC_COLLATE=POSIX per http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap07.html#tag_07_03_02_06 ...

I think that is the definitive definition, with everything outside of that table being unspecified.  Not something you'd want to mix with most characters outside of US-ASCII.  I think the `und-u-va-posix` name was invented by Go as part of their mad scheme to conflate language and locale golang/go#25340 ;).

[1]: https://github.com/containers/libpod/blame/v1.1.2/docs/podman.1.md#L46
[2]: https://github.com/containers/libpod/pull/2501#issuecomment-469903974

**Comment by mtrmac** [comment]
> > Even if it seems to be “we are doing the right thing to support different languages and cultures”, this is in effect “we have designed the system so that it breaks because you want your address book sorted correctly for your language, and your system administrator uses a different language”.
> 
> ... and the admin and user decided to use the same hook directories (high probability)
> and the admin decided to lead with characters that sort differently in different locales (low probability). I think that last mitigation is important.

Sure, this is all very much a corner case of the design, but the total effect is that we are doing so much extra work to get the corner case of the design _wrong_. Why would we? There’s really no benefit at all to imposing human collating preferences on the filenames, which are mostly comments on the _contents_ of those files; sorting comments on data in a list is not really something anyone ever asks for.

(English is not my native language; supporting Czech sorting in a directory named `hooks.d` is just… bizarre. Sure, there could well be systems where names are someone’s native language if at all possible… but we put `oci-register-machine.json` in there by default.)


OTOH, I do acknowledge that a fair amount of older shell scripts doing something like `for file in config.d/*; do …` would end up being `LC_COLLATE`-dependent, arguably without the authors ever realizing this.

> > ... it instead of (pretending to) support subtly-break-my-system alternatives.
> 
> I'm in favor of splitting on strings to get `en_US.UTF-8` -> `en_US` (which does parse correctly) and similar.

That’s even more code to get the corner case designed wrong. Not all that attractive to me.


> > That’s true, but being forced to maintain an independent configuration just because the software makes it hard to use the system-wide one is not optimal.
> 
> Implicit hook directories were [already deprecated in favor of `--hooks-dir`](https://github.com/containers/libpod/blame/v1.1.2/docs/podman.1.md#L46) (#1920) before this discussion. So I don't see this as a big loss.

Yes, it is a corner case. I know it is a corner case and most people will know to steer away from it.

> > I haven’t found anything definitive; meanwhile the POSIX definition of LC_COLLATE=POSIX per http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap07.html#tag_07_03_02_06 ...
> 
> I think that is the definitive definition, with everything outside of that table being unspecified.

The one thing it _does_ specify is that normalization must not be performed, so…

… I guess I’ll be testing that if I am ever that bored?

**Comment by wking** [comment]
> OTOH, I do acknowledge that a fair amount of older shell scripts doing something like for file in `config.d/*; do …` would end up being `LC_COLLATE`-dependent, arguably without the authors ever realizing this.

Ah, good point.  So to pull the Czech example from [here][1]:

```console
$ LC_COLLATE=cs_CZ.UTF-8 ls
baa  baá  báa  bab  báb  bac  bác  bač  báč
$ LC_COLLATE=en_US.UTF-8 ls
baa  báa  baá  bab  báb  bac  bác  bač  báč
```

Give those all `.json` suffixes and hook content, and, with the bugfix from #2550, `podman` will sort hooks in the same order as `ls`, which is good for low-surprise.  And as we keep affirming to each other, this probably doesn't matter anyway, because a hook directory full of `baá.json`, `báa.json`, etc., where locale matters is unlikely.  And again, I'm still fine with either respecting the user's configured collation locate or hard-coding `und-u-va-posix`.  I just prefer some kind of Unicode comparison over byte-for-byte comparison.

[1]: https://en.wikipedia.org/wiki/Alphabetical_order#Language-specific_conventions

**Comment by mtrmac** [comment]
> Give those all `.json` suffixes and hook content, and, with the bugfix from #2550, `podman` will sort hooks in the same order as `ls`, which is good for low-surprise.

Yet again, most C code is likely to use `strcmp` (or something similar to get ASCIIbetic order) and not `strcoll`, certainly all C code dealing with `*.d` that I’ve seen (notably https://github.com/systemd/systemd/blob/faab72d16b310c17be4b908cfe15eca122d16ae4/src/basic/conf-files.c#L175 ). And most of the shell consumers of `*.d`-like directories, like Fedora `initscripts`, have already been mostly phased out.

**Comment by wking** [comment]
> Yet again, most C code is likely to use `strcmp`...

I still think it would be a bug to use byte-by-byte sorting that ignored [Unicode normalization][1].  We should not sort differently for combining characters vs. composites.  And I'm still agnostic on `LC_COLLATE` vs. hard-coding `und-u-va-posix`.

[1]: http://www.unicode.org/reports/tr15/

**Comment by rhatdan** [comment]
@mtrmac @wking Whats the skoop on this one.  Have you guys settled on the best outcome?

**Comment by wking** [comment]
I'm still at my previous position.  @mtrmac, are you still committed to byte-sorting without Unicode normalization?

**Comment by mtrmac** [comment]
Most importantly, after #2550, we are no longer polluting the output with warnings. So, if no-one cares, just close this.
- I remain very convinced that using users’ locale to sort hooks is just bad design. At best, the argument for doing that is https://github.com/containers/libpod/pull/2501#issuecomment-469903974 , more or less (I’ve been trying go be fair but may have failed) “your configuration after the default install contains an explicit language setting, and that setting being present gives us the license to behave inconsistently, here is the inconsistent behavior you ‘asked for’”.
- Similarly (but weaker), using _any_ human collation order (i.e. one that can change with a software update) seems the wrong tradeoff WRT software testability and stability
- Once the language/collation order is decided, I don’t _really_ care about doing/not doing the Unicode normalization. It just seems pointless to me (nothing else in the codebase accounts for Unicode normalization, why would we do that only for hooks?), and rather expensive (in a quick comparison, this costs us 2.08 MiB in unstripped file size, or 1.65 MiB per `size(1)`.

**Comment by mtrmac** [comment]
> > > I haven’t found anything definitive; meanwhile the POSIX definition of LC_COLLATE=POSIX per http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap07.html#tag_07_03_02_06 ...
> > 
> > 
> > I think that is the definitive definition, with everything outside of that table being unspecified.
> 
> The one thing it _does_ specify is that normalization must not be performed, so…
> 
> … I guess I’ll be testing that if I am ever that bored?

\*sigh\* For the record, `und-u-va-posix` _is_ normalization aware, which makes it explicitly different from `LC_COLLATE=POSIX` (which ignore normalization, both on macOS and Linux).

```go
package main

// #include <locale.h>
// #include <stdlib.h>
// #include <string.h>
import "C"

import (
	"fmt"
	"sort"
	"strings"
	"unsafe"

	"golang.org/x/text/collate"
	"golang.org/x/text/language"
	"golang.org/x/text/unicode/norm"
)

var inputStrings = []string{
	"baa",
	"baá",
	"báa",
	"bab",
	"báb",
	"bac",
	"bác",
	"bač",
	"báč",
}

type input struct {
	value string
	norm  norm.Form
}

func newInput(s string, norm norm.Form) input {
	return input{
		value: norm.String(s),
		norm:  norm,
	}
}

var formName = map[norm.Form]string{
	norm.NFC: "NFC",
	norm.NFD: "NFD",
}

func printResults(name string, vals []input) {
	fmt.Printf("%s\n", name)
	fmt.Printf("%s\n", strings.Repeat("=", len(name)))
	fmt.Printf("%s\t%s\t%-14s\t%-14s\t%s\n", "Text", "Norm", "Raw", "NFC", "NFD")
	for _, v := range vals {
		fmt.Printf("%s\t%s\t%-14x\t%-14x\t%x\n", v.value, formName[v.norm], []byte(v.value), norm.NFC.Bytes([]byte(v.value)), norm.NFD.Bytes([]byte(v.value)))
	}

}

type inputsLister []input

func (il inputsLister) Len() int {
	return len(il)
}

func (il inputsLister) Swap(i, j int) {
	il[i], il[j] = il[j], il[i]
}

func (il inputsLister) Bytes(i int) []byte {
	return []byte(il[i].value)
}

func strcoll(a, b string) int {
	ca := C.CString(a)
	defer C.free(unsafe.Pointer(ca))
	cb := C.CString(b)
	defer C.free(unsafe.Pointer(cb))

	return int(C.strcoll(ca, cb))
}

func main() {
	vals := []input{}
	for _, s := range inputStrings {
		vals = append(vals, newInput(s, norm.NFC), newInput(s, norm.NFD))
	}

	sort.Slice(vals, func(i, j int) bool { return strings.ToLower(vals[i].value) < strings.ToLower(vals[j].value) })
	printResults("string < on strings.ToLower", vals)

	lang, err := language.Parse("und-u-va-posix")
	if err != nil {
		panic(err.Error())
	}
	collator := collate.New(lang, collate.IgnoreCase, collate.IgnoreWidth)
	collator.Sort(inputsLister(vals))
	printResults("collate(und-u-va-posix, IgnoreCase)", vals)

	empty := C.CString("")
	C.setlocale(C.LC_ALL, empty)
	check := C.setlocale(C.LC_ALL, nil)
	sort.Slice(vals, func(i, j int) bool { return strcoll(vals[i].value, vals[j].value) < 0 })
	printResults("strcoll "+C.GoString(check), vals)
}
```

**Comment by rhatdan** [comment]
Ok I am leaning towards agreeing with @mtrmac

**Comment by TomSweeneyRedHat** [comment]
@mtrmac still some test issues here.  Might not hurt to rebase.

**Comment by mtrmac** [comment]
Rebased, tests pass.

**Pull request merged**